File: core/src/main/java/javabot/dao/AdminDao.java
package javabot.dao;

import javabot.model.Admin;
import javabot.model.Config;
import javabot.model.EventType;
import javabot.model.OperationEvent;
import javabot.model.criteria.AdminCriteria;
import org.pircbotx.User;

import javax.inject.Inject;
import java.time.LocalDateTime;
import java.util.List;

public class AdminDao extends BaseDao<Admin> {
    @Inject
    private ConfigDao configDao;

    public AdminDao() {
        super(Admin.class);
    }

    @Override
    public List<Admin> findAll() {
        return ds.createQuery(Admin.class).order("userName").asList();
    }

    public boolean isAdmin(final User user) {
        return findAll().isEmpty() || getAdmin(user) != null;
    }

    public Admin getAdmin(final String ircName, final String hostName) {
        AdminCriteria adminCriteria = new AdminCriteria(ds);
        adminCriteria.ircName().equal(ircName);
        adminCriteria.hostName().equal(hostName);
        return adminCriteria.query().get();
    }

    public Admin getAdmin(final User user) {
        AdminCriteria adminCriteria = new AdminCriteria(ds);
        adminCriteria.ircName(user.getNick());
        return adminCriteria.query().get();
    }

    public Admin getAdminByEmailAddress(final String email) {
        AdminCriteria criteria = new AdminCriteria(ds);
        criteria.emailAddress(email);
        Admin admin = criteria.query().get();
        if(admin == null && count() == 0) {
            admin = new Admin();
            admin.setEmailAddress(email);
            admin.setBotOwner(true);
            save(admin);
        }
        return admin;
    }

    public Admin create(final String ircName, final String userName, final String hostName) {
        final Admin admin = new Admin();
        admin.setIrcName(ircName);
        admin.setEmailAddress(userName);
        admin.setHostName(hostName);
        admin.setUpdated(LocalDateTime.now());
        admin.setBotOwner(findAll().isEmpty());
        save(admin);

        return admin;
    }

    public void enableOperation(String name, Admin admin) {
        OperationEvent event = new OperationEvent();
        event.setOperation(name);
        event.setRequestedOn(LocalDateTime.now());
        event.setType(EventType.ADD);
        event.setRequestedBy(admin.getEmailAddress());
        save(event);
        Config config = configDao.get();
        config.getOperations().add(name);
        configDao.save(config);
    }

    public void disableOperation(String name, Admin admin) {
        OperationEvent event = new OperationEvent();
        event.setOperation(name);
        event.setRequestedOn(LocalDateTime.now());
        event.setType(EventType.DELETE);
        event.setRequestedBy(admin.getEmailAddress());
        save(event);
        Config config = configDao.get();
        config.getOperations().remove(name);
        configDao.save(config);
    }

    public long count() {
        return getQuery().countAll();
    }
}

File: src/main/kotlin/javabot/dao/AdminDao.kt
package javabot.dao

import dev.morphia.Datastore
import dev.morphia.query.FindOptions
import dev.morphia.query.Sort
import dev.morphia.query.filters.Filters.eq
import dev.morphia.query.filters.Filters.or
import javabot.model.Admin
import javabot.model.EventType
import javabot.model.JavabotUser
import javabot.model.OperationEvent
import javax.inject.Inject

class AdminDao @Inject constructor(ds: Datastore, var configDao: ConfigDao) :
    BaseDao<Admin>(ds, Admin::class.java) {
    override fun findAll(): List<Admin> {
        return ds.find(Admin::class.java)
            .iterator(FindOptions().sort(Sort.ascending("ircName")))
            .toList()
    }

    fun isAdmin(user: JavabotUser): Boolean = findAll().isEmpty() || getAdmin(user) != null

    fun getAdmin(ircName: String, hostName: String) =
        ds.find(Admin::class.java).filter(eq("ircName", ircName), eq("hostName", hostName)).first()

    fun getAdmin(user: JavabotUser) =
        ds.find(Admin::class.java)
            .filter(or(eq("ircName", user.nick), eq("emailAddress", user.userName)))
            .first()

    fun getAdminByEmailAddress(email: String) =
        ds.find(Admin::class.java).filter(eq("emailAddress", email)).first()

    fun create(ircName: String, userName: String, hostName: String): Admin {
        val admin = Admin(ircName, userName, hostName, findAll().isEmpty())

        save(admin)

        return admin
    }

    fun enableOperation(name: String, admin: Admin) {
        save(OperationEvent(admin.emailAddress, EventType.ADD, name))
        val config = configDao.get()
        config.operations.add(name)
        configDao.save(config)
    }

    fun disableOperation(name: String, admin: Admin) {
        save(OperationEvent(admin.emailAddress, EventType.DELETE, name))
        val config = configDao.get()
        config.operations.remove(name)
        configDao.save(config)
    }

    fun count(): Long {
        return getQuery().count()
    }
}
